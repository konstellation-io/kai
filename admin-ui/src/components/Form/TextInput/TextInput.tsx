import React, {
  useState,
  useEffect,
  MouseEvent,
  KeyboardEvent,
  FocusEvent,
  ChangeEvent,
  ReactElement
} from 'react';

import InputLabel from '../InputLabel/InputLabel';
import InputError from '../InputError/InputError';
import InputHelp from '../InputHelp/InputHelp';
import IconShow from '@material-ui/icons/RemoveRedEye';
import IconHide from '@material-ui/icons/RemoveRedEyeOutlined';
import { isFieldAnInteger } from '../check';

import cx from 'classnames';
import styles from './TextInput.module.scss';

type Props = {
  onSubmit?: Function;
  onChange?: Function;
  onBlur?: Function;
  placeholder?: string;
  label?: string;
  textArea?: boolean;
  height?: number;
  lockHorizontalGrowth?: boolean;
  limits?: {
    minWidth?: number;
    maxWidth?: number;
    minHeight?: number;
    maxHeight?: number;
  };
  error?: string;
  helpText?: string;
  showClearButton?: boolean;
  whiteColor?: boolean;
  onlyNumbers?: boolean;
  positive?: boolean;
  formValue?: string | number;
  customClassname?: string;
  hidden?: boolean;
  autoFocus?: boolean;
  maxLength?: number;
};

function TextInput({
  onSubmit = function() {},
  onChange = function() {},
  onBlur = function() {},
  placeholder = '',
  label = '',
  textArea = false,
  height = 40,
  lockHorizontalGrowth = false,
  limits = {},
  error = '',
  helpText = '',
  showClearButton = false,
  whiteColor = false,
  onlyNumbers = false,
  positive = false,
  formValue = '',
  customClassname = '',
  hidden = false,
  autoFocus = false,
  maxLength
}: Props) {
  const [value, setValue] = useState(formValue);
  const [isHidden, setIsHidden] = useState(hidden);

  useEffect(() => {
    setIsHidden(hidden);
  }, [hidden, setIsHidden]);

  useEffect(() => {
    setValue(formValue);
  }, [formValue, setValue]);

  function updateValue(newValue: string) {
    if (
      !onlyNumbers ||
      (onlyNumbers && isFieldAnInteger(newValue, positive).valid)
    ) {
      setValue(newValue);
      onChange(newValue);
    }
  }

  function onType(e: ChangeEvent<HTMLInputElement | HTMLTextAreaElement>) {
    const newValue = e.target.value;

    updateValue(newValue);
  }

  function onKeyPress(
    e: KeyboardEvent<HTMLInputElement | HTMLTextAreaElement>
  ) {
    if (e.which === 13 && !textArea) {
      // Enter key
      onSubmit(value);
    }
  }

  function onInputBlur(e: FocusEvent<HTMLInputElement | HTMLTextAreaElement>) {
    onBlur();
  }

  function toggleVisibility(e: MouseEvent<HTMLDivElement>) {
    setIsHidden(!isHidden);
  }

  const inputProps = {
    className: cx(styles.input, {
      [styles.error]: error !== '',
      [styles.lockHorizontalGrowth]: lockHorizontalGrowth
    }),
    value: value,
    type: isHidden ? 'password' : 'text',
    placeholder: placeholder,
    onChange: onType,
    onKeyPress: onKeyPress,
    onBlur: onInputBlur,
    style: { height },
    maxLength: maxLength
  };
  const inputElement: ReactElement =
    textArea && !isHidden ? (
      <textarea
        {...inputProps}
        data-testid="input"
        style={{ ...limits }}
        autoFocus={autoFocus}
      />
    ) : (
      <input {...inputProps} data-testid="input" autoFocus={autoFocus} />
    );
  const cleanButton =
    showClearButton && value !== '' ? (
      <div
        className={styles.clearButton}
        onClick={() => updateValue('')}
        data-testid="clear-button"
      >
        x
      </div>
    ) : (
      ''
    );
  const VisibilityIcon = isHidden ? IconShow : IconHide;
  const showEyeButton = hidden ? (
    <div
      className={cx(styles.eyeButton, {
        [styles.showClearButton]: showClearButton && value !== ''
      })}
      onClick={toggleVisibility}
    >
      <VisibilityIcon className="icon-small" />
    </div>
  ) : (
    ''
  );

  return (
    <div
      className={cx(styles.container, customClassname, {
        [styles.white]: whiteColor,
        [styles.hasClearButton]: showClearButton,
        [styles.hasEyeButton]: hidden
      })}
    >
      <InputLabel text={label} />
      {inputElement}
      {cleanButton}
      {showEyeButton}
      {error ? (
        <InputError message={error} />
      ) : (
        <InputHelp message={helpText} />
      )}
    </div>
  );
}

export default TextInput;
