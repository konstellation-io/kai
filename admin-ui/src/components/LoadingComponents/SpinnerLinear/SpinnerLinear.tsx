import React from 'react';

import Lottie from '../../Lottie/Lottie';
import animationData from './SpinnerLinear.json';

import styles from './SpinnerLinear.module.scss';

type Props = {
  size?: number;
};

function SpinnerLinear({ size = 125 }: Props = {}) {
  return (
    <div
      className={styles.loaderContainer}
      style={{ width: size, height: size }}
      data-testid="spinner"
    >
      <Lottie
        options={{ animationData }}
        width={size}
        height={size}
        segments={[0, 100]}
        forceSegments
      />
    </div>
  );
}

export default SpinnerLinear;
