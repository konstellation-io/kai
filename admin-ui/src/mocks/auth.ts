import { GET_USER_EMAIL } from '../components/Header/Header.graphql';

export const usernameMock = {
  request: {
    query: GET_USER_EMAIL
  },
  result: {
    data: {
      me: {
        email: 'user@konstellation.com'
      }
    }
  }
};

export const unauthorizedUsernameMock = {
  request: {
    query: GET_USER_EMAIL
  },
  result: {
    data: {
      me: {
        email: 'unauthorizedUser@konstellation.com'
      }
    }
  },
  error: new Error('Unauthorized user')
};
