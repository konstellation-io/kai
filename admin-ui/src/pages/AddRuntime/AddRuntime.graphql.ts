import gql from 'graphql-tag';
import { Runtime } from '../../graphql/models';

export interface AddRuntimeResponse {
  addRuntime: {
    errors?: string[];
    runtime: Runtime;
  };
}

export interface AddRuntimeVars {
  input: {
    name: string;
  };
}

export const ADD_RUNTIME = gql`
  mutation CreateRuntime($input: CreateRuntimeInput!) {
    createRuntime(input: $input) {
      errors {
        code
        message
      }
      runtime {
        name
      }
    }
  }
`;
