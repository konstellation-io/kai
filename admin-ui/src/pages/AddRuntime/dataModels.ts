import gql from 'graphql-tag';

export const ADD_RUNTIME = gql`
  mutation AddRuntime($name: String!) {
    addRuntime(name: $name) {
      success
      message
    }
  }
`;
