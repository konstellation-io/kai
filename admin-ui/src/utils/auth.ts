import Cookies from 'js-cookie';

export const JWT_COOKIE = '__jwt';

export const isUserAuthenticated = () => {
  const jwt = Cookies.get(JWT_COOKIE);

  return jwt;
};
