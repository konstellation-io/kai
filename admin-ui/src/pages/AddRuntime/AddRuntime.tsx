import React, { useEffect } from 'react';
import { useHistory } from 'react-router';
import useInput from '../../hooks/useInput';

import TextInput from '../../components/Form/TextInput/TextInput';
import Button from '../../components/Button/Button';
import * as CHECK from '../../components/Form/check';
import * as PAGES from '../../constants/routes';

import styles from './AddRuntime.module.scss';

import { useMutation } from '@apollo/react-hooks';
import {
  ADD_RUNTIME,
  AddRuntimeResponse,
  AddRuntimeVars
} from './AddRuntime.graphql';
import {
  GET_DASHBOARD,
  GetDashboardResponse
} from '../Dashboard/Dashboard.graphql';

function verifyRuntimeName(value: string) {
  return CHECK.getValidationError([
    CHECK.isFieldNotEmpty(value),
    CHECK.isFieldAnString(value)
  ]);
}

function AddRuntime() {
  const history = useHistory();
  const { value, isValid, onChange, error, setError } = useInput(
    '',
    verifyRuntimeName
  );
  const [addRuntime, { loading, error: mutationError }] = useMutation<
    AddRuntimeResponse,
    AddRuntimeVars
  >(ADD_RUNTIME, {
    onCompleted: onCompleteAddRuntime,
    update(cache, updateResult) {
      if (updateResult.data !== undefined && updateResult.data !== null) {
        const newRuntime = updateResult.data.createRuntime.runtime;
        const cacheResult = cache.readQuery<GetDashboardResponse>({
          query: GET_DASHBOARD
        });

        if (cacheResult !== null) {
          const { runtimes, alerts } = cacheResult;
          cache.writeQuery({
            query: GET_DASHBOARD,
            data: { runtimes: runtimes.concat([newRuntime]), alerts }
          });
        }
      }
    }
  });

  useEffect(() => {
    if (mutationError) {
      setError(mutationError.toString());
    }
  }, [mutationError, setError]);

  function onCompleteAddRuntime(updatedData: any) {
    // TODO: CHECK FOR API ERRORS
    console.log(`${value} runtime created`);
    history.push(PAGES.DASHBOARD);
  }

  function onSubmit() {
    if (isValid()) {
      const input = { name: value };
      addRuntime({ variables: { input } });
    }
  }

  return (
    <div className={styles.bg}>
      <div className={styles.grid}>
        <div className={styles.container}>
          <h1>Add Runtime</h1>
          <p className={styles.subtitle}>
            Cras quis nulla commodo, aliquam lectus sed, blandit augue. Cras
            ullamcorper bibendum bibendum.
          </p>
          <div className={styles.content}>
            <TextInput
              whiteColor
              label="runtime name"
              error={error}
              onChange={onChange}
              onSubmit={onSubmit}
            />
            <div className={styles.buttons}>
              <Button
                primary
                label="SAVE"
                onClick={onSubmit}
                loading={loading}
              />
              <Button label="CANCEL" to={PAGES.DASHBOARD} />
            </div>
          </div>
        </div>
      </div>
    </div>
  );
}

export default AddRuntime;
