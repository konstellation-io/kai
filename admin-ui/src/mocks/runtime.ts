import { GET_DASHBOARD } from '../pages/Dashboard/Dashboard.graphql';
import { ADD_RUNTIME } from '../pages/AddRuntime/AddRuntime.graphql';

export const dashboardMock = {
  request: {
    query: GET_DASHBOARD
  },
  result: {
    data: {
      dashboard: {
        runtimes: [
          {
            id: '00001',
            status: 'ACTIVE',
            name: 'Some Name',
            creationDate: '2018-01-02'
          },
          {
            id: '00002',
            status: 'ACTIVE',
            name: 'Some Other Name',
            creationDate: '2018-02-03'
          }
        ],
        alerts: [
          {
            id: 'id0',
            type: 'error',
            message: 'some message',
            runtime: { id: '00001' }
          },
          {
            id: 'id1',
            type: 'error',
            message: 'some message 2',
            runtime: { id: '00001' }
          }
        ]
      }
    }
  }
};

export const addRuntimeMock = {
  request: {
    query: ADD_RUNTIME,
    variables: { name: 'New Runtime' }
  },
  result: {
    data: {
      createRuntime: {
        success: true,
        message: 'some message'
      }
    }
  }
};
