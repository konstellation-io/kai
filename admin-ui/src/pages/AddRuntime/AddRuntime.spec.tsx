import React from 'react';
import AddRuntime from './AddRuntime';
import { MockedProvider } from '@apollo/react-testing';
import { usernameMock } from '../../mocks/auth';
import { addRuntimeMock, dashboardMock } from '../../mocks/runtime';
import {
  testid,
  label,
  mountApolloComponent
} from '../../utils/testUtilsEnzyme';
import TextInput from '../../components/Form/TextInput/TextInput';
import SpinnerLinear from '../../components/LoadingComponents/SpinnerLinear/SpinnerLinear';

const mocks = [addRuntimeMock, dashboardMock, usernameMock];

function Wrapper() {
  return (
    <MockedProvider mocks={mocks} addTypename={false}>
      <AddRuntime />
    </MockedProvider>
  );
}

jest.mock('react-router', () => ({
  useHistory: jest.fn()
}));

describe('AddRuntime', () => {
  it('matches snapshot', async () => {
    const wrapper = await mountApolloComponent(<Wrapper />);

    expect(wrapper).toMatchSnapshot();
  });

  test('show right texts', async () => {
    const wrapper = await mountApolloComponent(<Wrapper />);

    expect(wrapper.exists('.container h1')).toBeTruthy();
    expect(wrapper.exists(label('SAVE'))).toBeTruthy();
  });

  it('handles input changes', async () => {
    const wrapper = await mountApolloComponent(<Wrapper />);

    wrapper
      .find(TextInput)
      .find('input')
      .simulate('change', { target: { value: '' } });

    wrapper.find(label('SAVE')).simulate('click');

    expect(wrapper.exists(SpinnerLinear)).toBeFalsy();

    expect(wrapper.find(testid('error-message')).text()).not.toBe('');

    wrapper
      .find(TextInput)
      .find('input')
      .simulate('change', { target: { value: 'New Runtime' } });

    wrapper.find(label('SAVE')).simulate('click');
    expect(wrapper.exists(SpinnerLinear)).toBeTruthy();
  });
});
