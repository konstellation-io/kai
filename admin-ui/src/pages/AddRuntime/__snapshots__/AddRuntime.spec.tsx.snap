// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders AddRuntime without crashing 1`] = `
<div>
  <div
    class="bg"
  >
    <div
      class="grid"
    >
      <div
        class="container"
      >
        <h1>
          Add Runtime
        </h1>
        <p
          class="subtitle"
        >
          Cras quis nulla commodo, aliquam lectus sed, blandit augue. Cras ullamcorper bibendum bibendum.
        </p>
        <div
          class="content"
        >
          <div
            class="container white"
          >
            <label
              class="label"
              data-testid="label"
            >
              RUNTIME NAME
            </label>
            <input
              class="input"
              data-testid="input"
              placeholder=""
              style="height: 40px;"
              type="text"
              value=""
            />
            
            <div
              class="errorMessage"
              data-testid="error-message"
            />
          </div>
          <div
            class="buttons"
          >
            <div
              class="btn default middle primary"
              style="height: 40px; line-height: 40px;"
            >
              <span>
                SAVE
              </span>
            </div>
            <a
              data-testid="buttonLink"
              href="/dashboard"
            >
              <div
                class="btn default middle label"
                style="height: 40px; line-height: 40px;"
              >
                <span>
                  CANCEL
                </span>
              </div>
            </a>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
